File: app/lib/models/list_message.dart
// models/list_message.dart
class ListMessage {
  final int id;
  final int listId;
  final int userId;
  final String message;
  final MessageType messageType;
  final String? imageUrl;
  final bool isRead;
  final DateTime createdAt;
  final DateTime updatedAt;

  // User info (populated from API)
  final MessageUser? user;

  // Read status
  final int? readByCount;

  ListMessage({
    required this.id,
    required this.listId,
    required this.userId,
    required this.message,
    required this.messageType,
    this.imageUrl,
    this.isRead = false,
    required this.createdAt,
    required this.updatedAt,
    this.user,
    this.readByCount,
  });

  factory ListMessage.fromJson(Map<String, dynamic> json) {
    return ListMessage(
      id: json['id'] as int,
      listId: json['list_id'] as int,
      userId: json['user_id'] as int,
      message: json['message'] as String,
      messageType: MessageType.fromString(json['message_type'] as String? ?? 'text'),
      imageUrl: json['image_url'] as String?,
      isRead: json['is_read'] == 1 || json['is_read'] == true,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'] as String)
          : DateTime.now(),
      user: json['user'] != null
          ? MessageUser.fromJson(json['user'] as Map<String, dynamic>)
          : null,
      readByCount: json['read_by_count'] as int?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'list_id': listId,
      'user_id': userId,
      'message': message,
      'message_type': messageType.value,
      'image_url': imageUrl,
      'is_read': isRead,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      if (user != null) 'user': user!.toJson(),
      if (readByCount != null) 'read_by_count': readByCount,
    };
  }

  ListMessage copyWith({
    int? id,
    int? listId,
    int? userId,
    String? message,
    MessageType? messageType,
    String? imageUrl,
    bool? isRead,
    DateTime? createdAt,
    DateTime? updatedAt,
    MessageUser? user,
    int? readByCount,
  }) {
    return ListMessage(
      id: id ?? this.id,
      listId: listId ?? this.listId,
      userId: userId ?? this.userId,
      message: message ?? this.message,
      messageType: messageType ?? this.messageType,
      imageUrl: imageUrl ?? this.imageUrl,
      isRead: isRead ?? this.isRead,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      user: user ?? this.user,
      readByCount: readByCount ?? this.readByCount,
    );
  }

  bool isFromCurrentUser(int currentUserId) {
    return userId == currentUserId;
  }
}

/// Message type enum
enum MessageType {
  text('text'),
  image('image'),
  system('system');

  final String value;
  const MessageType(this.value);

  static MessageType fromString(String value) {
    switch (value.toLowerCase()) {
      case 'image':
        return MessageType.image;
      case 'system':
        return MessageType.system;
      case 'text':
      default:
        return MessageType.text;
    }
  }
}

/// User info in message
class MessageUser {
  final int id;
  final String name;
  final String email;

  MessageUser({
    required this.id,
    required this.name,
    required this.email,
  });

  factory MessageUser.fromJson(Map<String, dynamic> json) {
    return MessageUser(
      id: json['id'] as int,
      name: (json['name'] as String?) ?? 'Unknown User',
      email: (json['email'] as String?) ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'email': email,
    };
  }

  String get displayName {
    // Extract first name or use full name
    final parts = name.split(' ');
    return parts.isNotEmpty ? parts[0] : name;
  }

  String get initials {
    final parts = name.split(' ');
    if (parts.length >= 2) {
      return '${parts[0][0]}${parts[1][0]}'.toUpperCase();
    } else if (name.isNotEmpty) {
      return name.substring(0, 1).toUpperCase();
    }
    return '?';
  }
}

File: api/src/Models/ListMessage.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class ListMessage extends Model
{
    use SoftDeletes;

    protected $table = 'list_messages';
    protected $primaryKey = 'id';

    public $timestamps = true;

    protected $fillable = [
        'list_id',
        'user_id',
        'message',
        'message_type',
        'image_url',
        'is_read',
        'created_at',
        'updated_at',
        'deleted_at'
    ];

    protected $casts = [
        'is_read' => 'boolean',
        'created_at' => 'datetime',
        'updated_at' => 'datetime',
        'deleted_at' => 'datetime'
    ];

    protected $hidden = [
        'deleted_at'
    ];

    /**
     * Get the shopping list that owns the message.
     */
    public function shoppingList(): BelongsTo
    {
        return $this->belongsTo(ShoppingList::class, 'list_id');
    }

    /**
     * Get the user that created the message.
     */
    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    /**
     * Get the read status for this message.
     */
    public function readStatus(): HasMany
    {
        return $this->hasMany(MessageReadStatus::class, 'message_id');
    }

    /**
     * Check if a user has read this message.
     */
    public function isReadBy(int $userId): bool
    {
        return $this->readStatus()
            ->where('user_id', $userId)
            ->exists();
    }

    /**
     * Mark message as read by a user.
     */
    public function markAsReadBy(int $userId): void
    {
        MessageReadStatus::firstOrCreate([
            'message_id' => $this->id,
            'user_id' => $userId
        ]);
    }

    /**
     * Get unread count for a list by a user.
     */
    public static function getUnreadCountForList(int $listId, int $userId): int
    {
        return self::where('list_id', $listId)
            ->where('user_id', '!=', $userId) // Exclude own messages
            ->whereDoesntHave('readStatus', function ($query) use ($userId) {
                $query->where('user_id', $userId);
            })
            ->count();
    }

    /**
     * Format message for API response.
     */
    public function toArray(): array
    {
        $data = parent::toArray();

        // Add user info
        if ($this->relationLoaded('user') && $this->user) {
            $fullName = trim(($this->user->first_name ?? '') . ' ' . ($this->user->last_name ?? ''));
            if (empty($fullName)) {
                $fullName = $this->user->email ?? 'Unknown User';
            }

            $data['user'] = [
                'id' => $this->user->id,
                'name' => $fullName,
                'email' => $this->user->email ?? '',
            ];
        }

        // Add read count
        if ($this->relationLoaded('readStatus')) {
            $data['read_by_count'] = $this->readStatus->count();
        }

        return $data;
    }
}
